### ISSUE TITLE
CI VMs currently not happy

### ISSUE THREAD

**Issue by hzeller** [chipsalliance/verible#1663]
It looks like our runners are not coming up nicely
https://github.com/chipsalliance/verible/actions/runs/4057591116/jobs/6983523003

Some glitch or something that needs fixing ? Assigning to Adam for now as he is most familiar with the VM setup.

**Comment by AdamOlech** [comment]
I took a brief look and it seems to me that this could be correlated with the recent update to the Ubuntu Focal container on Docker Hub which took place merely [3 hours ago](https://hub.docker.com/layers/library/ubuntu/focal/images/sha256-d5d4814ffb155d588f10ec8926d9e1cd09b6d41a3110d2f42959e2fc37f6d0b4?context=explore).

I'm not sure what the exact cause of it could be (could be some API change or maybe rotten cache on our end as we do use [registry cache](https://github.com/antmicro/runner/blob/vm-runners/docker-registry-config.yml)).

For the time being I'd suggest to change the container either to a different Ubuntu version (e.g. 22) or to the Focal-compatible Debian Bullseye ([source](https://askubuntu.com/a/445496)).

In any case we'll investigate this error.

**Comment by AdamOlech** [comment]
We have investigated this issue thoroughly and arrived at a conclusion that this was caused by a mismatch of the `Content-Type` header reported by the registry and the actual manifest type that was being returned. 

To be precise, the registry reported that the MIME type was `application/vnd.oci.image.manifest.v1+json` while it was actually `application/vnd.docker.distribution.manifest.v2+json`. The difference between these two formats is negligible (practically none apart from different MIME type strings) but the libraries responsible for parsing manifests that were used Singularity (now called Apptainer) have fairly rigid constraints and will refuse to play nice in cases like these.

Although it seems like Ubuntu has fixed this bug upstream (i.e. the `latest` tags no longer point to faulty manifests), we have decided to implement some countermeasures to avoid downtime should this occur again:
* we migrated to Apptainer; it's simply a continuation of the Singularity project, more details on this "rename" can be found in the [official community announcement](https://apptainer.org/news/community-announcement-20211130/)
  * since we went with the most recent release, we benefit from more up-to-date versions of libraries for handling manifests and other Docker/OCI specific stuff
  * considering that it derives from the same codebase, it should be backwards compatible. Our tests have shown that this is the case, i.e. no changes in behavior have been observed. 
* we developed some patches on top of Apptainer to mitigate the content mismatch issue for this particular case (see [the patch](https://github.com/antmicro/buildroot/blob/2021.02.2-antmicro-main/package/apptainer/0001-Handle-Docker-manifest-V2-content-type-mismatch.patch)).

To summarize this: the issue no longer persists upstream but we are prepared for it nevertheless should it resurface.

@hzeller feel free to close the issue if you believe we're done here.

**Comment by hzeller** [comment]
Thanks for your thorough work on this and the detailed write-up, @AdamOlech !
Yes, I think this can be closed.
